### DISCUSSION TITLE
no decode delegate for this image format `JB2E'

### DISCUSSION THREAD

**Discussion by oavaldezi** [ImageMagick/ImageMagick#7847] [Help] [answered]
### ImageMagick version

7.1.1-41

### Operating system, version and so on

Windows 11

### Description

I'd like to convert some jb2e (JBIG2) files to JPG, but I get this error:

```
mogrify: unable to open module file 'C:\Program Files\ImageMagick-7.1.1-Q16-HDRI\modules\coders\IM_MOD_RL_JB2E_.dll': No such file or directory @ warning/module.c/GetMagickModulePath/678.

mogrify: no decode delegate for this image format `JB2E' @ error/constitute.c/ReadImage/746.
```

There is a "C:\Program Files\ImageMagick-7.1.1-Q16-HDRI\modules\coders\IM_MOD_RL_jbig_.dll" file on my system, but no IM_MOD_RL_jb2e_.dll.

I've renamed the files I wish to convert from *.jb2e to *.jbig, but that doesn't work, either:

```
mogrify: no decode delegate for this image format `JBIG' @ error/constitute.c/ReadImage/746.

mogrify: no decode delegate for this image format `' @ error/constitute.c/ReadImage/746.
```

Is it possible to convert JBIG2 files with Imagemagick ?

**Comment by fmw42** [comment]
What was your exact command line?

Did you use magick mogrify or just mogrify?  Have you tried just magick?

What do you get from 

`magick -version`

Does the list of delegates include jbig?

**Reply by oavaldezi** [reply]
Version: ImageMagick 7.1.1-41 Q16-HDRI x64 bbdcbf7:20241116 https://imagemagick.org

command line: magick mogrify -format jpg -quality 90 mgk-*.jb2e
 (I'm using magick mogrify because I want to batch process several JBIG2 files.)

These are included in the output from magick -list format
```
JBG* JBIG      ---   Joint Bi-level Image experts Group interchange format
JBIG* JBIG      ---   Joint Bi-level Image experts Group interchange format
```

I had to change the extension of the attached file to *.jpg so I could upload it. Please change the extension back to *.jb2e.
![mgk-003](https://github.com/user-attachments/assets/e7973092-3a75-4baf-8b49-68ff76f6f8ee)

**Comment by dlemstra** [comment]
Can you share your file?

**Reply by oavaldezi** [reply]
I had to change the extension of the attached file to *.jpg so I could upload it. Please change the extension back to *.jb2e.
![mgk-003](https://github.com/user-attachments/assets/e7973092-3a75-4baf-8b49-68ff76f6f8ee)

**Reply by fmw42** [reply]
Next time zip the file first, then you can upload as is

**Reply by oavaldezi** [reply]
Here is the zipped jb2e file.

Thank you for your replies.
[pdfimagesall-003.zip](https://github.com/user-attachments/files/18211706/pdfimagesall-003.zip)

**Reply by fmw42** [reply]
I am not an expert on this format, but your file does not look to me like a JBIG format file.  IM may only support JBIG and JBIG2(e).  An IM developer will need to comment.

**Reply by oavaldezi** [reply]
That's an interesting observation. It was extracted from a PDF file with poppler's `pdfimages -all`.

According to pdfimages' documentation, it will "_Write images in JBIG2 format as JBIG2 files instead of the default format. JBIG2 data in PDF is of the embedded type. The embedded type of JBIG2 has an optional separate file containing global data. The embedded data is written with the extension .jb2e and the global data (if available) will be written to the same image number with the extension .jb2g. The content of both these files is identical to the JBIG2 data in the PDF._"

In my case, `pdfimages -all` only produces a .jb2e file, without an accompanying .jb2g file.

**Comment by fmw42** [comment, marked as answer]
magick -list format

shows

      JBG* JBIG      rw+   Joint Bi-level Image experts Group interchange format (2.1)
     JBIG* JBIG      rw+   Joint Bi-level Image experts Group interchange format (2.1)

But not JBIG2.

**Reply by oavaldezi** [reply]
Can I add it to my Windows installation?

`magick -list compress` shows JBIG2

**Reply by fmw42** [reply]
I do not know. Is there a delegate for that?

See
https://en.wikipedia.org/wiki/JBIG
https://en.wikipedia.org/wiki/JBIG2

Can you tell pdfimages to export in another format?

**Reply by fmw42** [reply]
pdfimages supports TIFF or PNG output.

**Reply by oavaldezi** [reply]
Yes. It can export the .jb2e images to .png. But I don't think they are of type Bilevel, as the .jb2e images are.

**Comment by fmw42** [comment]
PNG and TIFF support bilevel, I believe

**Reply by oavaldezi** [reply]
Yes, I've tested a few operations on pdfimages PNG output, and I can retain their small size with `magick -type Bilevel`.

Thank you, you've helped me find a working solution.
